File: python-dist/src/qntm/charter/crypto.py
"""The charter's strict prime-order Ed25519 profile, shared with QSP."""

from ..ed25519 import is_valid_ed25519_public_key, verify_ed25519_signature
from .json import CharterError


def validate_charter_public_key(public_key: bytes) -> None:
    if type(public_key) is not bytes or len(public_key) != 32:
        raise CharterError("Expected 32-byte Ed25519 charter key")
    if not is_valid_ed25519_public_key(public_key):
        raise CharterError("Invalid prime-order Ed25519 charter key")


def verify_charter_signature(public_key: bytes, message: bytes, signature: bytes) -> bool:
    validate_charter_public_key(public_key)
    if type(signature) is not bytes:
        return False
    return verify_ed25519_signature(public_key, message, signature)

File: client/src/charter/crypto.ts
import { isValidEd25519PublicKey, verifyEd25519Signature } from '../crypto/ed25519.js';

/** Charter keys use canonical, non-identity points in the prime-order subgroup. */
export function validateCharterPublicKey(publicKey: Uint8Array): void {
  if (!isValidEd25519PublicKey(publicKey)) {
    throw new Error('Invalid prime-order Ed25519 charter key');
  }
}

export function verifyCharterSignature(publicKey: Uint8Array, message: Uint8Array, signature: Uint8Array): boolean {
  validateCharterPublicKey(publicKey);
  return verifyEd25519Signature(publicKey, message, signature);
}
